### PR TITLE
Bug 2020622: Create steps for AWS UPI

### PR DESCRIPTION
**Pull request by rna-afk**
Switching AWS UPI tests from templates to steps. Added 2 different steps for UPI and Proxy UPI installation.

### CODE REVIEW

**Comment by jstuever** [comment]
This will need to incorporate #23268 once it merges.

**Comment by jstuever** [comment]
Or, maybe this needs to incorporate that now... I don't think it will merge.

File: ci-operator/step-registry/upi/deprovision/aws/upi-deprovision-aws-commands.sh
Comment (staebler) [inline ci-operator/step-registry/upi/deprovision/aws/upi-deprovision-aws-commands.sh:43]: This is failing due to a non-ascii horizontal ellipsis in the console output. I don't think we need to block this PR to fix this, but we should at least file a card to fix it later.

File: ci-operator/step-registry/upi/deprovision/aws/upi-deprovision-aws-commands.sh
Comment (staebler) [inline ci-operator/step-registry/upi/deprovision/aws/upi-deprovision-aws-commands.sh:53]: It looks like we are getting a couple stacks that are failing to delete.
```
Waiter StackDeleteComplete failed: Waiter encountered a terminal failure state: For expression "Stacks[].StackStatus" we matched expected path: "DELETE_FAILED" at least once

Waiter StackDeleteComplete failed: Waiter encountered a terminal failure state: For expression "Stacks[].StackStatus" we matched expected path: "DELETE_FAILED" at least once
```
I wonder if there may be anything helpful that we can get from `aws cloudformation describe-stacks --stack-name "${CLUSTER_NAME}-${STACK_SUFFIX}` to provide more information in the output about which stack failed and why.

**Comment by staebler** [comment]
Looks like the ipi-deprovision is still not working, too.
```
Deprovisioning cluster ...
Skipping: /tmp/secret/metadata.json not found.
```

File: ci-operator/step-registry/upi/deprovision/aws/check-destroy/upi-deprovision-aws-check-destroy-commands.sh
Comment (jstuever) [inline ci-operator/step-registry/upi/deprovision/aws/check-destroy/upi-deprovision-aws-check-destroy-commands.sh:1]: I believe this behavior should be part of `upi-deprovision-aws.sh` instead of having to duplicate the effort here. I expect the reason it is here is because there are things that need to happen between the first call of `upi-deprovision-aws.sh` and when this is actually called. I believe `upi-aws-post-chain` can be changed to remove this.

File: ci-operator/step-registry/upi/aws/post/upi-aws-post-chain.yaml
Comment (jstuever) [inline ci-operator/step-registry/upi/aws/post/upi-aws-post-chain.yaml:7]: I expect reodrering these will remove the need to call `upi-deprovision-aws-check-destroy` and enable that functionality to be instead included in `upi-deprovision-aws`. By first calling `ipi-deprovision-deprovosion`, the cluster should mostly be cleaned up (and therefore unable to create new things). Then calling `upi-deprovision-aws` with the full funcionality in the `check-destroy` variant will finish things up. This will remove the need for `upi-deprovision-aws-check-destroy` altogether.

**Review by jstuever — CHANGES_REQUESTED** [review]
There are a few situations where you are duplicating effort. I recommend combining these to reduce differences in upi-aws and upi-aws-proxy from cropping up slowly over time.

File: ci-operator/step-registry/upi/install/aws/upi-install-aws-commands.sh
Comment (rna-afk) [reply ci-operator/step-registry/upi/install/aws/upi-install-aws-commands.sh:57]: That gives me an error with shellcheck. (Declare and assign separately)

File: ci-operator/step-registry/upi/deprovision/aws/check-destroy/upi-deprovision-aws-check-destroy-commands.sh
Comment (rna-afk) [reply ci-operator/step-registry/upi/deprovision/aws/check-destroy/upi-deprovision-aws-check-destroy-commands.sh:1]: We had a problem with some kind of cyclic dependency error where the vpc and the security groups associated with the vpc were not getting deleted. I will try to run it one more time to see if I face that error and replace this if necessary

File: ci-operator/step-registry/upi/aws/pre/upi-aws-pre-chain.yaml
Comment (rna-afk) [reply ci-operator/step-registry/upi/aws/pre/upi-aws-pre-chain.yaml:7]: I was thinking creating another folder for the chain and then calling just the ref(upi-install-aws) and the ipi-install-times-collection would have more files than just add two lines here but I see your concern where this may be a problem in the future with respect to modularity? I will do the former.

Comment (rna-afk) [reply ci-operator/step-registry/upi/aws/pre/upi-aws-pre-chain.yaml:7]: The former involves a lot of extra folders(for the upi and proxy steps) that I think are redundant and I feel it's probably best to just add it here? Thoughts?

File: ci-operator/step-registry/upi/aws/post/upi-aws-post-chain.yaml
Comment (rna-afk) [reply ci-operator/step-registry/upi/aws/post/upi-aws-post-chain.yaml]: I reordered it and now I get this cyclic dependency error here. 
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/22317/rehearse-22317-pull-ci-openshift-installer-master-e2e-aws-upi/1485976419107344384

Comment (jstuever) [reply ci-operator/step-registry/upi/aws/post/upi-aws-post-chain.yaml:7]: I didn't look at this in too much detail; but it looks like possibly there is an object that's not getting tagged properly in the UPI template and therefore not deleted by the installer binary.

File: ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh
Comment (staebler) [inline ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh:74]: See https://bugzilla.redhat.com/show_bug.cgi?id=2047799. The preferred approach with jq is to pass in environment variables via the --arg flag.

Comment (rna-afk) [reply ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh:74]: I think it was more of a quoting issue? This code does not have the single quotes but the code in templates does.
https://github.com/openshift/release/commit/3c29ef1ea4988141d8c97cd0db55e8dad554ba09#diff-ba2cb3cdc0f179f886cf0bf9167c1251e87209e78d18f8bff73cec1463afd02aR1314

Comment (staebler) [reply ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh:74]: Yes, the issue in the template was quoting. But I do think it is better to use the --arg flag and avoid the quoting quagmire altogether. Just a nit: Feel free to keep what you have.
```suggestion
  RHCOS_AMI="$(jq -r --arg region "$AWS_REGION" '.architectures.x86_64.images.aws.regions[$region].image' coreos.json)"
```

File: ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh
Comment (jstuever) [inline ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh:76]: Does this have the same quoting issue as line 74?

**Comment by jstuever** [comment]
I had one concern about RHCOS_IMAGE quoting on the else statement. Otherwise, I think we should go ahead and merge this after we get the e2e-aws-upi and e2e-aws-upi-proxy passing. The work to combine the upi check destroy step into the upi destroy step can be added to our backlog as tech debt (even if we continue working to resolve that).

File: ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh
Comment (rna-afk) [reply ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh]: Good point. It doesn't but it looks inconsistent with the previous line so I'll change it

**Comment by rna-afk** [comment]
e2e-aws-upi test has passed. 

/hold cancel

**Pull request merged**
